File: algo_problemset/cpp/q109_convert_sorted_list_to_bst.cpp
/*
 * Description:
 *
 *
 *
 * @author: Ricky Chang
*/

#include <iostream>
#include <vector>

using namespace std;


// Definition for singly-linked list.
struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};


// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {

private:

    bool canAdvanceFastPtr(ListNode* fastPtr) {
        return (fastPtr != NULL) && (fastPtr->next != NULL);
    }

    ListNode* getBeforeMidPoint(ListNode* node) {
        ListNode* slowPtr = node;
        ListNode* fastPtr = node;

        while (true) {
            fastPtr = fastPtr->next->next;
            if (!canAdvanceFastPtr(fastPtr)) { break; }
            slowPtr = slowPtr->next;
        }

        return slowPtr;
    }

    TreeNode* splitList(ListNode* node) {
        ListNode* beforeMidNode;
        TreeNode* rootNode;

        if (node == NULL) {
            return NULL;
        } else if (node->next == NULL) {
            rootNode = new TreeNode(node->val);
            return rootNode;
        }

        beforeMidNode = getBeforeMidPoint(node);
        rootNode = new TreeNode(beforeMidNode->next->val);

        rootNode->right = splitList(beforeMidNode->next->next);
        // disconnect the list from the middle
        beforeMidNode->next = NULL;
        rootNode->left = splitList(node);

        return rootNode;
    }

public:

    TreeNode* sortedListToBST(ListNode* head) {
        return splitList(head);
    }
};


ListNode* createList(vector<int> v) {
    ListNode* nodeStart = new ListNode(v[0]);
    ListNode* nodePtr = nodeStart;

    int n = v.size();
    for (int i = 1; i < n; ++i) {
        nodePtr->next = new ListNode(v[i]);
        nodePtr = nodePtr->next;
    }
    return nodeStart;
}

void printInOrder(TreeNode* root) {
    if (root == NULL) { return; }
    printInOrder(root->left);
    cout << root->val << ' ';
    printInOrder(root->right);
}


int main() {

    Solution sol;
    ListNode* list = createList({1, 3, 5, 8, 9, 10, 12, 18});
    TreeNode* root = sol.sortedListToBST(list);

    printInOrder(root);
    cout << endl;

    return 0;
}

File: algo_problemset/python/q109_convert_sorted_list_to_bst.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:

    def sortedListToBST(self, head: ListNode) -> TreeNode:
        n = self.getLinkedListSize(head)
        nodePtr = head

        def recursiveTransform(start: int, end: int) -> TreeNode:
            """ Return the root of the subtree coming from
                start to end of the ListNode
            """
            nonlocal nodePtr

            diff = end - start
            if diff == 0:
                return None
            if diff == 1:
                subRootNode = toTreeNode(nodePtr)
                nodePtr = nodePtr.next
                return subRootNode

            mid = (start + end) // 2
            leftChild = recursiveTransform(start, mid)
            rootNode = toTreeNode(nodePtr)
            rootNode.left = leftChild

            nodePtr = nodePtr.next
            rootNode.right = recursiveTransform(mid+1, end)

            return rootNode

        rootNode = recursiveTransform(0, n)
        return rootNode

    def getLinkedListSize(self, head: ListNode) -> int:
        cnt = 0
        while head is not None:
            head = head.next
            cnt += 1
        return cnt


def toTreeNode(node: ListNode) -> TreeNode:
    return TreeNode(node.val)
